daskms/parallel_table.py

```diff
@@ -0,0 +1,86 @@
+import logging
+import threading
+from pyrap.tables import table as Table
+from weakref import finalize
+
+
+log = logging.getLogger(__name__)
+
+
+def _parallel_table_finalizer(_cached_tables):
+
+    for table in _cached_tables.values():
+        table.close()
+
+
+class ParallelTable(Table):
```

Comment:
https://www.developer.com/design/encapsulation-vs-inheritance/

Reply:
I am trying to grasp this, but I am still struggling. The problem I see is that failing to inherit from `pyrap.tables.table` means that the ParallelTable object will not in fact appear to be a table i.e. `self._table_future = table = ex.impl.submit(factory, *args, **kwargs)` will not be a future pointing at a table. What I could see working is defining a `ParralelTableProxy` which simply inherits from `TableProxy` but defines its own metaclass which modifies the behaviour of get* operations. Currently, `ParallelTable` is itself a table (i.e. you can do something like `pt.taql(query, ParallelTable)`) in addition to simply rerouting get* operation through a cache. In other words, there will always be one extra copy open - the "root" table which gets used for unmodified methods. I will take a swing at the `ParallelTableProxy` idea.

Reply:
Actually, would you mind clarifying what the problem is with this approach? Currently, all the `ParallelTable` does is override some inherited methods of `pyrap.tables.table` prior (!!) to the being embedded in a `TableProxy`. This just means that the proxy proxies these special methods, rather than those on the base class. This yields a really simple solution as subsequent operations proceed as normal. The `ParallelTable` is a table, and supports all `pyrap.tables.table` methods, and will have all the relevant multiton patterns applied inside the `TableProxy`. I have tried creating a `ParallelTableProxy`, but that becomes difficult as one needs to access the cache inside get* methods.

Reply:
I have implemented a `ParallelTableProxy`. At present it segfaults, but I know why. The problem stems from the fact that `getter_wrapper` in `reads.py` calls methods directly on the underlying table object, rather than via the `TableProxy`. This is problematic as `TableProxy.method` may not be the same as `pyrap.tables.table.method`. This is precisely where my current segfaults come from, as `getter_wrapper` calls the non-threadsafe get* functions on the underlying table.